src/CoreWCF.Kafka/src/CoreWCF/Channels/KafkaTransportPump.cs

```diff
@@ -274,4 +294,72 @@ public void Dispose()
             _mres?.Dispose();
         }
     }
+
+    internal class TopicPartitionOffsetTracker
+    {
+        private readonly ConcurrentDictionary<TopicPartition, SortedList<ConsumeResult<Null, byte[]>, bool>> _topicPartitions = new();
```

Comment:
`SortedList<T, U>` is O(n^2) for insert and remove. Replace with a Fibonacci heap. See `TimerQueue` in `IOThreadTimer`